{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Modified by Marco Fumasoni from \"plot_tumor_depth_all_chromosome.py\" of unknown origin\n",
    "\n",
    "# This version of the script subtracts the 'normal' depth ratio (called anc) to the 'tumor' depth ratio (called evo) \n",
    "# from a varscan .copynumber file. In that way naturally occurring repeated sequences are buffered. \n",
    "# This result in more linear chromosome visualization and highlights only the fragments that have a real change in \n",
    "# read depth across the genome. To facilitate the identification the script uses a color gradient where red means gain \n",
    "# of copies and blue means loss. \n",
    "# The script generate 3 pdf files, one for the anc genome, one for the evo genome and one for the evo genome normalized to\n",
    "# to the ancestor\n",
    "\n",
    "# NOTE\n",
    "# The code is made available for transparency reasons. At present, it is not intended to be readily usable on different datasets. \n",
    "# Also, it was not annotated and compiled to be user-friendly. Please, contact me privately for any inquiry related to the code usage.\n",
    "# I will maintain this code with improved versions as soon as they are developed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Anaconda2\\lib\\site-packages\\ipykernel\\__main__.py:88: RuntimeWarning: invalid value encountered in divide\n",
      "C:\\Anaconda2\\lib\\site-packages\\ipykernel\\__main__.py:100: RuntimeWarning: invalid value encountered in divide\n",
      "C:\\Anaconda2\\lib\\site-packages\\ipykernel\\__main__.py:112: RuntimeWarning: invalid value encountered in divide\n",
      "C:\\Anaconda2\\lib\\site-packages\\ipykernel\\__main__.py:252: RuntimeWarning: invalid value encountered in greater\n",
      "C:\\Anaconda2\\lib\\site-packages\\ipykernel\\__main__.py:253: RuntimeWarning: invalid value encountered in less\n",
      "C:\\Anaconda2\\lib\\site-packages\\ipykernel\\__main__.py:254: RuntimeWarning: invalid value encountered in greater_equal\n",
      "C:\\Anaconda2\\lib\\site-packages\\ipykernel\\__main__.py:254: RuntimeWarning: invalid value encountered in less_equal\n",
      "C:\\Anaconda2\\lib\\site-packages\\matplotlib\\pyplot.py:516: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n",
      "  max_open_warning, RuntimeWarning)\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "from matplotlib import cm\n",
    "import glob, os\n",
    "import sys\n",
    "#import seaborn as sns\n",
    "from scipy.stats import mode\n",
    "from matplotlib.patches import Rectangle\n",
    "\n",
    "\n",
    "# Find the .copynumber files in the folder and generate 3 pdf for each file\n",
    "os.chdir(\"../clones\")\n",
    "for file in glob.glob(\"*.copynumber\"):\n",
    "    # load the varscan dataset\n",
    "    datafn=file\n",
    "\n",
    "    # Read in the start position and log2_ratio for each window, saving each chromosome separately\n",
    "    # Initialize a dictionary to hold the data, arranged by chromosome\n",
    "    chrom_dict = {}\n",
    "    with open(datafn) as datafile:\n",
    "        # The first line of the file is the header line. Store in variable `header,' which we will ignore.\n",
    "        header = datafile.readline()\n",
    "        # Read the file one line at a time\n",
    "        for line in datafile:\n",
    "            # The first position in the line gives the chromosome\n",
    "            chrom = line.split()[0]\n",
    "            # If we don't already have an entry for the chromosome in our dictionary, create an entry.\n",
    "            # The entry consists of two empty lists, one will hold the start positions and one will hold the coverage\n",
    "            if chrom not in chrom_dict:\n",
    "                chrom_dict[chrom] = ([], [],[],[])\n",
    "            # Store the start position, coverage_ratio,normal_depth,tumor_depth in their respective lists\n",
    "            chrom_dict[chrom][0].append(int(line.split()[1]))\n",
    "            chrom_dict[chrom][1].append(float(line.split()[-2]))\n",
    "            chrom_dict[chrom][2].append(float(line.split()[-4]))\n",
    "            chrom_dict[chrom][3].append(float(line.split()[-3]))\n",
    "\n",
    "    genome_list_anc = []\n",
    "    genome_list_evo = []\n",
    "\n",
    "    W = 1000 #set the size of the window\n",
    "    \n",
    "    #store anc depth for entire genome\n",
    "    #calculate the median of normal depth across the whole genome\n",
    "    for value in chrom_dict.itervalues():\n",
    "        genome_list_anc.extend(value[2])\n",
    "\n",
    "    genome_array_anc = np.array(genome_list_anc)\n",
    "    genome_median_anc = np.median(genome_array_anc)\n",
    "\n",
    "    # store evo depth for entire genome\n",
    "    #calculate the median of tumor depth across the whole genome\n",
    "    for value in chrom_dict.itervalues():\n",
    "        genome_list_evo.extend(value[3])\n",
    "\n",
    "    genome_array_evo = np.array(genome_list_evo)\n",
    "    genome_median_evo = np.median(genome_array_evo)\n",
    "\n",
    "    # Loop over the different chromosomes, plot the coverage, and save the figure\n",
    "\n",
    "    chrom_list = sorted(chrom_dict.keys())\n",
    "    x_vals = []\n",
    "    y_vals = []\n",
    "    x_vals_anc = []\n",
    "    y_vals_anc = []\n",
    "    x_vals_evo = []\n",
    "    y_vals_evo = []\n",
    "    for chrom in chrom_list:\n",
    "        # Get the start positions and coverage ratio from our dictionary.\n",
    "        # Convert the data from lists to numpy arrays (better for mathematical operations and plotting)\n",
    "        start = np.array(chrom_dict[chrom][0])\n",
    "        cov_ratio = np.array(chrom_dict[chrom][1])\n",
    "        anc_depth = np.array(chrom_dict[chrom][2])\n",
    "        evo_depth = np.array(chrom_dict[chrom][3])\n",
    "\n",
    "        # Normalize by median of genome depth to control for difference sequencing depths\n",
    "        normalized_depth_anc = anc_depth / genome_median_anc\n",
    "        normalized_depth_evo = evo_depth / genome_median_evo\n",
    "        normalized_depth = (normalized_depth_evo) - (normalized_depth_anc)\n",
    "        \n",
    "        # Calculate smoothed coverage over different window sizes for the ancestor\n",
    "\n",
    "        smoothed = np.zeros((start[-1] / W) + 1)\n",
    "        n_bins = np.zeros_like(smoothed)\n",
    "        for i in range(len(normalized_depth)):\n",
    "            smoothed[start[i]/W] += normalized_depth_anc[i]\n",
    "            n_bins[start[i]/W] += 1\n",
    "        smoothed /= n_bins\n",
    "\n",
    "        x_vals_anc.append(W*np.arange(len(smoothed)))\n",
    "        y_vals_anc.append(smoothed)\n",
    "        \n",
    "        # Calculate smoothed coverage over different window sizes for the evolved line\n",
    "       \n",
    "        smoothed = np.zeros((start[-1] / W) + 1)\n",
    "        n_bins = np.zeros_like(smoothed)\n",
    "        for i in range(len(normalized_depth)):\n",
    "            smoothed[start[i]/W] += normalized_depth_evo[i]\n",
    "            n_bins[start[i]/W] += 1\n",
    "        smoothed /= n_bins\n",
    "\n",
    "        x_vals_evo.append(W*np.arange(len(smoothed)))\n",
    "        y_vals_evo.append(smoothed)\n",
    "        \n",
    "        # Calculate smoothed coverage over different window sizes for the normalized evo genome\n",
    "        \n",
    "        smoothed = np.zeros((start[-1] / W) + 1)\n",
    "        n_bins = np.zeros_like(smoothed)\n",
    "        for i in range(len(normalized_depth)):\n",
    "            smoothed[start[i]/W] += normalized_depth[i]\n",
    "            n_bins[start[i]/W] += 1\n",
    "        smoothed /= n_bins\n",
    "\n",
    "        x_vals.append(W*np.arange(len(smoothed)))\n",
    "        y_vals.append(smoothed)\n",
    "\n",
    "\n",
    "    # create a 2D array for the coordinates of the 16 centromeres\n",
    "    centromeres = [[0 for x in range(2)] for y in range(16)] \n",
    "    # populate the array with the coordinates +- the windows size (to be sure to visualize the centromere)\n",
    "    centromeres[0][0]= 151465 - W\n",
    "    centromeres[0][1]= 151582 + W\n",
    "    centromeres[1][0]= 238207 - W\n",
    "    centromeres[1][1]= 238323 + W\n",
    "    centromeres[2][0]= 114385 - W\n",
    "    centromeres[2][1]= 114501 + W\n",
    "    centromeres[3][0]= 449711 - W\n",
    "    centromeres[3][1]= 449711 + W\n",
    "    centromeres[4][0]= 151987 - W\n",
    "    centromeres[4][1]= 152104 + W\n",
    "    centromeres[5][0]= 148510 - W\n",
    "    centromeres[5][1]= 148627 + W\n",
    "    centromeres[6][0]= 496920 - W\n",
    "    centromeres[6][1]= 497038 + W\n",
    "    centromeres[7][0]= 105586 - W\n",
    "    centromeres[7][1]= 105703 + W\n",
    "    centromeres[8][0]= 355629 - W\n",
    "    centromeres[8][1]= 355745 + W\n",
    "    centromeres[9][0]= 436307 - W\n",
    "    centromeres[9][1]= 436425 + W\n",
    "    centromeres[10][0]= 440129 - W\n",
    "    centromeres[10][1]= 440246 + W\n",
    "    centromeres[11][0]= 150828 - W\n",
    "    centromeres[11][1]= 150947 + W\n",
    "    centromeres[12][0]= 268031 - W\n",
    "    centromeres[12][1]= 268149 + W\n",
    "    centromeres[13][0]= 628758 - W\n",
    "    centromeres[13][1]= 628875 + W\n",
    "    centromeres[14][0]= 326584 - W\n",
    "    centromeres[14][1]= 326702 + W\n",
    "    centromeres[15][0]= 555957 - W\n",
    "    centromeres[15][1]= 556073 + W\n",
    "\n",
    "    #get the names of the two genomes compared\n",
    "    anc_text = file.partition(\"_vs_\")[0]\n",
    "    evo_text = file.partition(\"_vs_\")[2]\n",
    "    \n",
    "    #Open a pdf for the anc figure\n",
    "    fig_fn = anc_text + '.pdf'\n",
    "    pdf_pages = PdfPages(fig_fn)\n",
    "    \n",
    "    #Plot the figure for the ancestor \n",
    "    #These are to make the y scale of all panels the same\n",
    "    ymin = -1 \n",
    "    ymax = 4 \n",
    "    \n",
    "    fig, ax = plt.subplots(4,4, sharey ='row')\n",
    "    fig.text(0.5, 0.01, 'chromosome coordinate', ha='center', va='center',fontsize=8)\n",
    "    fig.text(0.01, 0.5, 'normalized read depth', ha='center', va='center', rotation='vertical',fontsize=8)\n",
    "    for i, a in enumerate(ax.flatten()):\n",
    "        a.plot(x_vals_anc[i], y_vals_anc[i], alpha=0.3)\n",
    "        a.scatter(x_vals_anc[i], y_vals_anc[i], marker='.', s=10, linewidth='0', vmin=-1, vmax=1, zorder=10)\n",
    "        a.set_ylim([ymin, ymax])\n",
    "        a.yaxis.set_ticks(np.arange(ymin,(ymax+0.5),0.5))\n",
    "        xmax=max(x_vals[i])\n",
    "        a.set_xlim(-15000, xmax+15000)\n",
    "        a.xaxis.set_tick_params(labelsize=3)\n",
    "        a.yaxis.set_tick_params(labelsize=3)\n",
    "        a.set_title('chromosome '+ str(i+1), fontsize=8)  \n",
    "    fig.tight_layout()\n",
    "    pdf_pages.savefig(fig)\n",
    "    #plt.show()\n",
    "    #Save the pdf file\n",
    "    pdf_pages.close()\n",
    "    \n",
    "    #Open a pdf for the evo figure\n",
    "    fig_fn = evo_text + '.pdf'\n",
    "    pdf_pages = PdfPages(fig_fn)\n",
    "    \n",
    "    #Plot the figure for the evolved line\n",
    "    # These are to make the y scale of all panels the same\n",
    "    ymin = -1 \n",
    "    ymax = 4 \n",
    "    \n",
    "    fig, ax = plt.subplots(4,4, sharey ='row')\n",
    "    fig.text(0.5, 0.01, 'chromosome coordinate', ha='center', va='center',fontsize=8)\n",
    "    fig.text(0.01, 0.5, 'normalized read depth', ha='center', va='center', rotation='vertical',fontsize=8)\n",
    "    for i, a in enumerate(ax.flatten()):\n",
    "        a.plot(x_vals_evo[i], y_vals_evo[i], alpha=0.3)\n",
    "        a.scatter(x_vals_evo[i], y_vals_evo[i], marker='.', s=10, linewidth='0', vmin=-1, vmax=1, zorder=10)\n",
    "        a.set_ylim([ymin, ymax])\n",
    "        a.yaxis.set_ticks(np.arange(ymin,(ymax+0.5),0.5))\n",
    "        xmax=max(x_vals[i])\n",
    "        a.set_xlim(-15000, xmax+15000)\n",
    "        a.xaxis.set_tick_params(labelsize=3)\n",
    "        a.yaxis.set_tick_params(labelsize=3)\n",
    "        a.set_title('chromosome '+ str(i+1), fontsize=8)  \n",
    "    fig.tight_layout()\n",
    "    pdf_pages.savefig(fig)\n",
    "    #plt.show()\n",
    "    #Save the pdf file\n",
    "    pdf_pages.close()\n",
    "    \n",
    "    # create two 2D arrays to store the centromere coordinates and the relative read depth\n",
    "    x_vals_cen = []\n",
    "    y_vals_cen = []\n",
    "    # find the data for each chromosome of the normalized genome\n",
    "    for i in range(0,16):\n",
    "        cen_coordinate = ((x_vals[i] >= centromeres[i][0]) & (x_vals[i] <= centromeres[i][1]))\n",
    "        chromosome = x_vals[i]\n",
    "        chr_reads = y_vals[i]\n",
    "        x_vals_cen.append([])\n",
    "        y_vals_cen.append([])\n",
    "        x_vals_cen[i].append(chromosome[cen_coordinate])\n",
    "        y_vals_cen[i].append(chr_reads[cen_coordinate])\n",
    "        x_vals[i] = chromosome[~cen_coordinate]\n",
    "        y_vals[i] = chr_reads[~cen_coordinate]\n",
    "\n",
    "    #Open a pdf for the normalized evo genome figure\n",
    "    fig_fn = file + '_black.pdf'\n",
    "    pdf_pages = PdfPages(fig_fn)\n",
    "    \n",
    "    #Plot the figure of the normalized genome with a color gradient\n",
    "    fig, ax = plt.subplots(4,4,dpi=900)\n",
    "    fig.text(0.5, 0.01, 'chromosome coordinate', ha='center', va='center',fontsize=8)\n",
    "    fig.text(0.01, 0.5, 'normalized read depth', ha='center', va='center', rotation='vertical',fontsize=8)\n",
    "    \n",
    "    #x_max_temp=0\n",
    "    #x_max=0\n",
    "    #for i in range (0,len(x_vals)):\n",
    "    #    x_max_temp = max(x_vals[i])\n",
    "    #    x_max = max([x_max, x_max_temp])\n",
    "    \n",
    "    ymax=3.5\n",
    "    ymin=-1\n",
    "    \n",
    "    for i, a in enumerate(ax.flatten()):\n",
    "        cp= y_vals[i]\n",
    "        \n",
    "        \n",
    "        x_vals[i]=x_vals[i]/1000\n",
    "        red= [x_vals[i][np.where(y_vals[i]>0.5)], y_vals[i][np.where(y_vals[i]>0.5)]]\n",
    "        blue= [x_vals[i][np.where(y_vals[i]<-0.5)], y_vals[i][np.where(y_vals[i]<-0.5)]]\n",
    "        black= [x_vals[i][np.where((y_vals[i]>=-0.5) & (y_vals[i]<=0.5))], y_vals[i][np.where((y_vals[i]>=-0.5) & (y_vals[i]<=0.5))]]\n",
    "        \n",
    "        a.plot(x_vals[i], y_vals[i], color='black', alpha=0.3)\n",
    "        #a.scatter(x_vals[i], y_vals[i], marker='.', s=10, linewidth='0', vmin=-1, vmax=1, color='black', alpha=0.4, zorder=10)\n",
    "        a.scatter(red[0],red[1], marker='.', s=10, linewidth='0', vmin=-1, vmax=1, alpha=0.7, color='black', zorder=10)\n",
    "        a.scatter(blue[0],blue[1], marker='.', s=10, linewidth='0', vmin=-1, vmax=1, alpha=0.7, color='black', zorder=10)\n",
    "        a.scatter(black[0],black[1], marker='.', s=10, linewidth='0', vmin=-1, vmax=1, alpha=0.4, color='black', zorder=10)\n",
    "        cen_avx=np.nanmean(x_vals_cen[i])\n",
    "        cen_avy=np.nanmean(y_vals_cen[i])\n",
    "        a.plot(cen_avx, cen_avy, 'ko', alpha=1, ms=2, zorder=20) \n",
    "        a.add_patch(Rectangle((cen_avx, cen_avy-1), (centromeres[i][1]-centromeres[i][0]) , 2, fill=None, alpha=0.4,zorder=20))\n",
    "        a.set_ylim([ymin, ymax])\n",
    "        a.yaxis.set_ticks(np.arange(ymin,(ymax+0.5),0.5))\n",
    "        x_max=max(x_vals[i])\n",
    "        a.set_xlim(-15, x_max+15)\n",
    "        a.xaxis.set_tick_params(labelsize=5, length=2)\n",
    "        a.yaxis.set_tick_params(labelsize=5, length=2)\n",
    "        a.set_title('chromosome '+ str(i+1), fontsize=8)  \n",
    "    fig.tight_layout()\n",
    "    pdf_pages.savefig(fig)\n",
    "    #plt.show()\n",
    "\n",
    "    #Save the pdf file\n",
    "    pdf_pages.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
